File: androidApp/src/main/java/com/luisma/conexiones/android/features/game/GameViewModel.kt
package com.luisma.conexiones.android.features.game

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewModelScope
import androidx.lifecycle.viewmodel.viewModelFactory
import com.luisma.conexiones.android.router.RoutePayload
import com.luisma.conexiones.android.router.RouterService
import com.luisma.conexiones.android.router.Routes
import com.luisma.conexiones.models.BasicScreenState
import com.luisma.conexiones.models.game.GameState
import com.luisma.conexiones.models.game.GameTileAnimationType
import com.luisma.conexiones.models.game.GameWordAnimationType
import com.luisma.conexiones.services.INumbUtilsService
import com.luisma.conexiones.services.IUserProfileService
import com.luisma.conexiones.services.game.GamePlayService
import com.luisma.conexiones.services.game.GameSelectionService
import com.luisma.conexiones.services.game.IGameAnimationService
import com.luisma.conexiones.services.gameAnimationService
import com.luisma.conexiones.services.gamePlayService
import com.luisma.conexiones.services.gameSelectionService
import com.luisma.conexiones.services.numbUtilsService
import com.luisma.conexiones.services.userProfileService
import kotlinx.coroutines.async
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

class GameViewModel(
    private val gamePlayService: GamePlayService,
    private val gameSelectionService: GameSelectionService,
    private val userProfileService: IUserProfileService,
    private val routerService: RouterService,
    private val numbUtilsService: INumbUtilsService,
    private val gameAnimationService: IGameAnimationService
) : ViewModel() {

    companion object {
        val Factory: ViewModelProvider.Factory = viewModelFactory {
            addInitializer(GameViewModel::class) {
                GameViewModel(
                    gamePlayService = gamePlayService(),
                    userProfileService = userProfileService(),
                    routerService = RouterService,
                    gameSelectionService = gameSelectionService(),
                    numbUtilsService = numbUtilsService(),
                    gameAnimationService = gameAnimationService()
                )
            }
        }
    }

    private val _state = MutableStateFlow(GameViewState.initial())
    val state = _state.asStateFlow()

    fun sendEvent(event: GameViewEvents) {
        when (event) {
            is GameViewEvents.OnCreate -> onCrate(gameId = event.gameId)
            is GameViewEvents.SelectWord -> selectWord(col = event.col, row = event.row)
            is GameViewEvents.DismissWordAnimation -> dismissWordAnimation(
                col = event.col,
                row = event.row
            )
            is GameViewEvents.DismissTileAnimation -> dismissTileAnimation(row = event.row)
            is GameViewEvents.DismissLivesAnimation -> dismissLivesAnimation()
            is GameViewEvents.DismissDoneSignAnimation -> dismissDoneSignAnimation()
            is GameViewEvents.GoToNextLevel -> goToNextLevel(nextLevelId = event.nextLevelId)
            GameViewEvents.MixGame -> mixGame()
            GameViewEvents.ClearSelection -> clearSelection()
            GameViewEvents.OnBack -> onBack()
            GameViewEvents.Submit -> submit()
            GameViewEvents.MoreLives -> {}// TODO:
        }
    }

    private fun onCrate(gameId: Int) {
        if (gameId == _state.value.gameId || gameId == -1) {
            return
        }
        startGame(gameId = gameId)
    }

    private fun startGame(gameId: Int) {
        viewModelScope.launch {
            val gameDef = async { gamePlayService.getGame(gameId = gameId) }
            val livesDef = async { userProfileService.getLives() }
            val game = gameDef.await()
            val lives = livesDef.await()
            _state.update {
                it.copy(
                    gameId = gameId,
                    screenState = BasicScreenState.Success,
                    gameData = game.gameData,
                    gridRowsState = game.gridRowsState,
                    lives = lives,
                    livesEarnedOnDone = game.livesEarnedOnDone,
                    nextGameId = game.nextGameId,
                    showOnDoneSignAnimation = game.showOnDoneSignAnimation,
                    showOnDoneLivesAnimation = game.showOnDoneLivesAnimation
                )
            }
        }
    }

    private fun mixGame() {
        viewModelScope.launch {
            val gameResponse = gamePlayService.mixGame(
                gameId = _state.value.gameId,
                gameDistribution = _state.value.gameData.gameDistribution,
                currentSelection = _state.value.currentSelection,
            )
            _state.update {
                it.copy(
                    gameData = gameResponse.gameData,
                    gridRowsState = gameResponse.gridRowsState,
                    currentSelection = gameResponse.selection
                )
            }
        }
    }

    private fun onBack() {
        viewModelScope.launch {
            routerService.setRoute(
                RoutePayload(
                    route = Routes.BackWithRefresh,
                    payload = numbUtilsService.getRandomInt().toString()
                )
            )
        }
    }

    private fun selectWord(col: Int, row: Int) {
        val result = gameSelectionService.select(
            gridRowState = _state.value.gridRowsState,
            currentSelection = _state.value.currentSelection,
            newColumn = col,
            newRow = row
        )

        _state.update {
            it.copy(
                gridRowsState = result.gridRowState,
                currentSelection = result.newSelection,
                selectFailed = false
            )
        }
    }

    private fun clearSelection() {
        val result = gameSelectionService.clearSelection(
            gridRowState = _state.value.gridRowsState,
            currentSelection = _state.value.currentSelection,
        )

        if (result == null) {
            return
        }

        _state.update {
            it.copy(
                gridRowsState = result.gridRowState,
                currentSelection = result.newSelection,
            )
        }
    }


    private fun submit() {
        viewModelScope.launch {
            val response = gamePlayService.submitSelection(
                gameId = _state.value.gameId,
                gameDistribution = _state.value.gameData.gameDistribution,
                currentSelection = _state.value.currentSelection
            )

            var livesEarnedOnDone = _state.value.livesEarnedOnDone
            var lives = _state.value.lives

            if (response.gameData.gameState == GameState.Win ||
                response.gameData.gameState == GameState.Lost
            ) {
                livesEarnedOnDone = response.livesEarnedOnDone
                lives = response.totalLivesAfterOnDone
            }

            _state.update {
                it.copy(
                    gameData = response.gameData,
                    gridRowsState = response.gridRowsState,
                    currentSelection = response.selection,
                    selectFailed = !response.rowGuessedOnSubmit,
                    livesEarnedOnDone = livesEarnedOnDone,
                    lives = lives,
                    nextGameId = response.nextGameId,
                    notSolvedRowsOnLost = response.notSolvedRowsOnLost,
                )
            }
        }
    }

    private fun goToNextLevel(nextLevelId: Int) {
        _state.update {
            GameViewState.initial()
        }

        startGame(gameId = nextLevelId)
    }

    private fun dismissWordAnimation(
        col: Int,
        row: Int,
    ) {
        val resp = gameAnimationService.toggleWordAnimation(
            gridRowState = _state.value.gridRowsState,
            column = col,
            row = row,
            animationType = GameWordAnimationType.NoAnimation
        )
        _state.update {
            it.copy(
                gridRowsState = resp
            )
        }
    }

    private fun dismissTileAnimation(
        row: Int,
    ) {
        val resp = gameAnimationService.toggleTilesAnimation(
            gridRowState = _state.value.gridRowsState,
            rows = listOf(row),
            animationType = GameTileAnimationType.NoAnimation
        )
        _state.update {
            it.copy(
                gridRowsState = resp
            )
        }
    }


    private fun dismissLivesAnimation() {
        if(!_state.value.showOnDoneLivesAnimation) return

        _state.update {
            it.copy(showOnDoneLivesAnimation = false)
        }
    }
    private fun dismissDoneSignAnimation() {
        if(!_state.value.showOnDoneSignAnimation) return

        _state.update {
            it.copy(showOnDoneSignAnimation = false)
        }
    }

}

File: iosApp/iosApp/Features/Game/GameViewModel.swift
//
//  GameViewModel.swift
//  iosApp
//
//  Created by Luis Manuel García Fernández on 10/10/24.
//  Copyright © 2024 orgName. All rights reserved.
//

import shared
import SwiftUI

class GameViewModel: ObservableObject {
    @Published public private(set) var state: GameViewState

    private let gamePlayService: GamePlayService
    private let gameSelectionService: GameSelectionService
    private let userProfileService: IUserProfileService
    private let numbUtilsService: INumbUtilsService
    private let gameAnimationService: IGameAnimationService
    private let router = Router.instance

    init(
        state: GameViewState = .init(),
        gamePlayService: GamePlayService = ServicesLocatorKt.gamePlayService(),
        gameSelectionService: GameSelectionService = ServicesLocatorKt.gameSelectionService(),
        userProfileService: IUserProfileService = ServicesLocatorKt.userProfileService(),
        numbUtilsService: INumbUtilsService = ServicesLocatorKt.numbUtilsService(),
        gameAnimationService: IGameAnimationService = ServicesLocatorKt.gameAnimationService()
    ) {
        self.state = state
        self.gamePlayService = gamePlayService
        self.gameSelectionService = gameSelectionService
        self.userProfileService = userProfileService
        self.numbUtilsService = numbUtilsService
        self.gameAnimationService = gameAnimationService
    }

    func sendEvent(event: GameViewEvents) {
        switch event {
        case .onAppear(gameId: let gameId):
            onAppear(gameId: gameId)
        case .selectWord(column: let column, row: let row):
            selectWord(column: column, row: row)
        case .dismissAnimation(column: let column, row: let row):
            dismissAnimation(column: column, row: row)
        case .dismissTileAnimation(row: let row):
            dissmissTileAnimation(row: row)
        case .dismissOnDoneAnimation:
            dismissOnDoneAnimation()
        case .dismissLivesAnimation:
            dismissLivesAnimation()
        case .goToNextLevel(nextLevel: let nextLevel):
            goToNextLevel(nextLevel: nextLevel)
        case .mixGame:
            mixGame()
        case .clearSelection:
            clearSelection()
        case .onBack:
            onBack()
        case .submit:
            submit()
        case .moreLives:
            moreLives()
        }
    }

    private func mapGridRowStateFromKt(_ kotlinGridRow: [KotlinInt: GameGridRowState]) -> [Int: GameGridRowState] {
        var gridRowsState: [Int: GameGridRowState] = [:]
        for idx in 0 ..< kotlinGridRow.count {
            gridRowsState[idx] = kotlinGridRow[KotlinInt(integerLiteral: idx)]
        }
        return gridRowsState
    }

    private func mapGridRowState2Kt(_ kotlinGridRow: [Int: GameGridRowState]) -> [KotlinInt: GameGridRowState] {
        var gridRowsState: [KotlinInt: GameGridRowState] = [:]
        for idx in 0 ..< kotlinGridRow.count {
            gridRowsState[KotlinInt(integerLiteral: idx)] = kotlinGridRow[idx]
        }
        return gridRowsState
    }

    private func onAppear(gameId: Int) {
        guard gameId != state.gameId || gameId != -1 else {
            return
        }

        startGame(gameId: gameId)
    }

    private func startGame(gameId: Int) {
        Task {
            async let gameResp = gamePlayService.getGame(gameId: Int32(gameId))
            async let livesResp = userProfileService.getLives()

            let game = try await gameResp
            let lives = try await livesResp

            await MainActor.run {
                state.gameId = gameId
                state.screenState = .success
                state.gameData = game.gameData
                state.gridRowsState = mapGridRowStateFromKt(game.gridRowsState)
                state.lives = Int(truncating: lives)
                state.livesEarnedOnDone = Int(game.livesEarnedOnDone)
                state.nextGameId = Int(game.nextGameId)
            }
        }
    }

    private func mixGame() {
        Task {
            let game = try await gamePlayService.mixGame(
                gameId: Int32(state.gameId),
                gameDistribution: state.gameData.gameDistribution,
                currentSelection: state.currentSelection
            )
            await MainActor.run {
                state.gameData = game.gameData
                state.gridRowsState = mapGridRowStateFromKt(game.gridRowsState)
                state.currentSelection = game.selection
            }
        }
    }

    private func selectWord(column: Int, row: Int) {
        let result = gameSelectionService.select(
            gridRowState: mapGridRowState2Kt(state.gridRowsState),
            currentSelection: state.currentSelection,
            newColumn: Int32(column),
            newRow: Int32(row)
        )
        Task {
            await MainActor.run {
                state.gridRowsState = mapGridRowStateFromKt(result.gridRowState)
                state.currentSelection = result.newSelection
                state.selectFailed = false
            }
        }
    }

    private func clearSelection() {
        let result = gameSelectionService.clearSelection(
            gridRowState: mapGridRowState2Kt(state.gridRowsState),
            currentSelection: state.currentSelection
        )

        Task {
            await MainActor.run {
                if let result = result {
                    state.gridRowsState = mapGridRowStateFromKt(result.gridRowState)
                    state.currentSelection = result.newSelection
                }
            }
        }
    }

    private func submit() {
        Task {
            let response = try await gamePlayService.submitSelection(
                gameId: Int32(state.gameId),
                gameDistribution: state.gameData.gameDistribution,
                currentSelection: state.currentSelection
            )

            await MainActor.run {
                var livesEarnedOnDone = state.livesEarnedOnDone
                var lives = state.lives

                if response.gameData.gameState == GameState.win ||
                    response.gameData.gameState == GameState.lost
                {
                    livesEarnedOnDone = Int(response.livesEarnedOnDone)
                    lives = Int(response.totalLivesAfterOnDone)
                }

                var notSolvedRowsOnLost: [Int: Int] = [:]
                response.notSolvedRowsOnLostIdx.forEach { key in
                    let keyInt = Int(truncating: key)
                    notSolvedRowsOnLost[keyInt] = Int(
                        truncating: response.notSolvedRowsOnLost[KotlinInt(integerLiteral: keyInt)] ?? 0
                    )
                }
                
                let gameData = response.gameData
                
                state.gameData = gameData
                state.gridRowsState = mapGridRowStateFromKt(response.gridRowsState)
                state.currentSelection = response.selection
                state.selectFailed = !response.rowGuessedOnSubmit
                state.livesEarnedOnDone = livesEarnedOnDone
                state.lives = lives
                state.nextGameId = Int(response.nextGameId)
                state.notSolvedRowsOnLost = notSolvedRowsOnLost
                state.showOnDoneSignAnimation = gameData.gameState == .win || gameData.gameState == .lost
                state.showOnDoneLivesAnimation = gameData.gameState == .win
            }
        }
    }

    private func goToNextLevel(nextLevel: Int) {
        state = .init()

        startGame(gameId: nextLevel)
    }

    func dismissAnimation(column: Int, row: Int) {
        let gridRowsState = gameAnimationService.toggleWordAnimation(
            gridRowState: mapGridRowState2Kt(state.gridRowsState),
            column: Int32(column),
            row: Int32(row),
            animationType: .noanimation
        )

        state.gridRowsState = mapGridRowStateFromKt(gridRowsState)
    }

    func dissmissTileAnimation(row: Int) {
        let gridRowsState = gameAnimationService.toggleTilesAnimation(
            gridRowState: mapGridRowState2Kt(state.gridRowsState),
            rows: [KotlinInt(integerLiteral: row)],
            animationType: .noanimation
        )

        state.gridRowsState = mapGridRowStateFromKt(gridRowsState)
    }

    func dismissOnDoneAnimation() {
        if !state.showOnDoneSignAnimation {
            return
        }

        state.showOnDoneSignAnimation = false
    }

    func dismissLivesAnimation() {
        if !state.showOnDoneLivesAnimation {
            return
        }

        state.showOnDoneLivesAnimation = false
    }

    private func onBack() {
        Task {
            await router.pop()
        }
    }

    private func moreLives() {
        Task {
            await router.goTo(page: .MockAdd)
        }
    }
}
